src/tools/illink/src/linker/Linker/RuntimeInterfaceImplementation.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Mono.Cecil;
+
+namespace Mono.Linker
+{
+	/// <summary>
+	/// Represents an implementation of an interface on a type that may be directly on the type or on a base type or implemented interface.
+	/// This type is considered to implement the interface at runtime, even though the interface may not be directly on the type.
+	/// </summary>
+	/// <remarks>
+	/// This type should be used for marking, but should NOT be used to check if a runtime interface implementation has been marked.
+	/// This type represents the most direct way an interface may be implemented, but it may be implemented in a less direct way that is not represented here.
+	/// You should check all possible implementation 'paths' to determine if an interface is implemented, for example <see cref="MarkStep.IsInterfaceImplementationMarkedRecursively"/>.
+	/// </remarks>
+	internal sealed class RuntimeInterfaceImplementation
+	{
+		/// <summary>
+		/// The type that implements <see cref="RuntimeInterfaceImplementation.InflatedInterfaceType"/>.
+		/// </summary>
+		public TypeDefinition Implementor { get; }
+
+		/// <summary>
+		/// The path of .interfaceimpl on <see cref="RuntimeInterfaceImplementation.Implementor"/> or a base type that terminates with <see cref="RuntimeInterfaceImplementation.InflatedInterfaceType"/>.
+		/// </summary>
+		public ImmutableArray<InterfaceImplementation> InterfaceImplementation { get; }
```

Comment:
Yeah, I think you're right that this shouldn't be implements/inherits tree order with only one .impl chain, and I think there are scenarios where we need to track at least more than one implementation.

```C#
class C : II, IJ, IK
{ }
interface II : IJ, IK
{ void M() {} }
interface IJ : IK
{ void M() {} }
interface IK
{ void M(); }
```

If we only mark the explicit interfaceImpls, `II`'s impl of `IJ` may be removed, which means that `II.M()` would no longer be the most specific implementation.

Thank you for helping me think this through.

Reply:
Ah, that sample should look more like below. In the example, C should implement II, IJ, and IK, but use the default interface method from II since it is the more specific implementation. However, if we only mark the shortest .impl chain, we mark 'C impls II, IK', 'II impls IK', but not necessarily 'II impls IJ', though that's necessary to know that II.M() is the most specific method.

```C#
class C : II, IJ, IK
{ }

interface II : IJ, IK
{ void IK.M() { Console.WriteLine("II"); } }

interface IJ : IK
{ void IK.M() { Console.WriteLine("IJ"); } }

interface IK
{ void M(); }
```

Reply:
Sorry, that was a poor explanation of the scenario. Using the same code as above, this is the scenario I'm thinking of.

```C#
class C : II, IJ, IK
{ }

interface II : IJ, IK
{ void IK.M() { Console.WriteLine("II"); } }

interface IJ : IK
{ void IK.M() { Console.WriteLine("IJ"); } }

interface IK
{ void M(); }
```

PreLink:
```mermaid
flowchart
    C --> II
    C --> IJ
    C --> IK
    II --> IJ
    II --> IK
    IJ --> IK
```
`II` is the most specific implementation because it inherits from both `IJ` and `IK` so it will provide the DIM.

If:
1. We mark "C implements IK": we'll mark the explicit .interfaceImpl, as well as the .interfaceImpl for "C implements II" and "II implements IK", and `II.M()` since `II` provides a DIM for `IK.M()` that C needs.
2. We mark "C implements IJ"
3. We mark `IJ.M()` and "IJ implements IK"

Then we've marked all but the relationship between `II` and `IJ`

```mermaid
flowchart
    C -->|1| II
    C -->|2| IJ
    C -->|1| IK
    II --> IJ
    II -->|1| IK
    IJ -->|3| IK
```

Linked:
```mermaid
flowchart
    C --> II
    C --> IJ
    C --> IK
    II --> IK
    IJ --> IK
```

In this scenario, both `II` and `IJ` implement `IK`, but the .interfaceImpl relationship between them is lost: there is no longer a "most specific implementation" for the method, and an exception would be thrown at runtime.

I don't think following the inherits/implements tree would solve this scenario. We could try to be smart about it, but I think the best way would be to mark all the possible interface chains between the interface and implementor.